File: assignment 7/c programming/p2.c
// Stack implementation in C using linked list
#include <stdio.h>
#include <stdlib.h>
#define MAX 100  // Maximum size of the stack
#include <limits.h>
#include <stdbool.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};
struct Node* top = NULL; // Initialize top of stack
int currentSize = 0; // Current size of the stack
// Function to check if the stack is full
bool isFull() {
    return currentSize >= MAX;
}
// Function to check if the stack is empty
bool isEmpty() {
    return top == NULL;
}
// Function to push an element onto the stack
void push(int value) {
    if (isFull()) {
        printf("Stack Overflow\n");
        return;
    }
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = top;
    top = newNode;
    currentSize++;
    printf("%d pushed to stack\n", value);
}
// Function to pop an element from the stack
void pop() {
    if (isEmpty()) {
        printf("Stack Underflow\n");
        return;
    }
    struct Node* temp = top;
    int poppedValue = top->data;
    top = top->next;
    free(temp);
    currentSize--;
    printf("%d popped from stack\n", poppedValue);
}

File: assignment 7/python/p2.py
# Stack implementation using a linked list
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def is_empty(self):
        return self.top is None

    def push(self, item):
        new_node = Node(item)
        new_node.next = self.top
        self.top = new_node

    def pop(self):
        if self.is_empty():
            raise IndexError("pop from empty stack")
        popped_node = self.top
        self.top = self.top.next
        return popped_node.data

    def peek(self):
        if self.is_empty():
            raise IndexError("peek from empty stack")
        return self.top.data

    def size(self):
        current = self.top
        count = 0
        while current:
            count += 1
            current = current.next
        return count
    
    def display(self):
        if self.is_empty():
            print("Stack is empty")
        else:
            print("Stack elements:")
            current = self.top
            while current:
                print(current.data)
                current = current.next
# Example usage
if __name__ == "__main__":
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.display()  # Output: 3, 2, 1
    print("Top element:", stack.peek())  # Output: 3
    print("Stack size:", stack.size())  # Output: 3
    print("Popped element:", stack.pop())  # Output: 3
    stack.display()  # Output: 2, 1
